site/src/pages/tutorials/thrift/blog/implement-create.mdx

```diff
@@ -0,0 +1,160 @@
+---
+menuTitle: "Implement CREATE"
+order: 4
+category: thrift
+tags:
+  - server
+level: basic
+type: step
+---
+
+# Implementing CREATE operation
+
+In the previous step, we defined empty service methods.
+In this step, we'll fill in one of the service methods to create a blog post and again, try making a call to a service method.
+
+<TutorialSteps current={4} />
+
+## What you need
+
+You need to have the following obtained from previous steps:
+
+- [Java code generated](/tutorials/thrift/blog/create-service#2-compile-the-thrift-file)
+- Main.java
+- BlogServiceImpl.java
+- BlogClient.java
+
+## 1. Implement server-side
+
+Let's implement a service method to create a blog post.
+
+1. In the `BlogServiceImpl` class, create an ID generator to issue temporary blog post IDs and a map to contain blog posts.
+  ```java filename=BlogServiceImpl.java
+  import java.util.Map;
+  import java.util.concurrent.ConcurrentHashMap;
+  import java.util.concurrent.atomic.AtomicInteger;
+
+  public class BlogServiceImpl implements BlogService.AsyncIface {
+    private final AtomicInteger idGenerator = new AtomicInteger();
+    private final Map<Integer, BlogPost> blogPosts = new ConcurrentHashMap<>();
+    ...
+  }
+  ```
+2. In the `createBlogPost()` method, create a `BlogPost` object with a generated ID and request parameters.
+  ```java filename=BlogServiceImpl.java
+  import java.time.Instant;
+  ...
+  @Override
+  public void createBlogPost(CreateBlogPostRequest request, AsyncMethodCallback<BlogPost> resultHandler)
+          throws TException {
+    final int id = idGenerator.getAndIncrement();
+    final Instant now = Instant.now();
+    final BlogPost blogPost = new BlogPost()
+            .setId(id)
+            .setTitle(request.getTitle())
+            .setContent(request.getContent())
+            .setModifiedAt(now.toEpochMilli())
+            .setCreatedAt(now.toEpochMilli());
+  }
+  ```
+2. In the `createBlogPost()` method, save the post information in the `blogPosts` map and return the information of the blog post created to the `resultHandler`.
+  ```java filename=BlogServiceImpl.java
+  @Override
+  public void createBlogPost(CreateBlogPostRequest request, AsyncMethodCallback<BlogPost> resultHandler)
+          throws TException {
+    ...
+    blogPosts.put(id, blogPost);
+    final BlogPost stored = blogPost;
+    resultHandler.onComplete(stored);
+  }
+  ```
+
+## 2. Implement client-side
+
+In the `BlogClient` class, add a method to send a request to create a blog post.
+
+```java filename=BlogClient.java
+import example.armeria.blog.thrift.BlogPost;
+import example.armeria.blog.thrift.CreateBlogPostRequest;
+...
+BlogPost createBlogPost(String title, String content) throws TException {
+  final CreateBlogPostRequest request =
+          new CreateBlogPostRequest().setTitle(title)
+                                     .setContent(content);
+  return blogService.createBlogPost(request);
+}
+```
+
+## 3. Set up for testing
```

Comment:
I'm happy to contribute more! 
I'm not sure how to modularize the testing at the moment, though 🤔 
I think more detailed suggestions would help me understand the how and develop ideas further.

Reply:
For instance, I was thinking we could add an extra step which describes how testing works in armeria.

It could describe:
- Adding the `armeria-junit5` module
- How to spin up a test server using `ServerExtension`
```
    @RegisterExtension
    static final ServerExtension server = ...
```
- How to call the server using a client
```
server.webClient()
server.httpUri()
```
- (Optional) How the tutorials use `@TestMethodOrder(OrderAnnotation.class)`

We're planning on migrating the other tutorials (rest, grpc) to also use tests (like done in this tutorial).
By modularizing, I thought the migration could be smoother (either by referencing or copying the section on `Testing in Armeria`).

I also found some references on how tests work for other web frameworks. I don't think it has to be as detailed as below, but I was thinking of something similar to below.
- https://spring.io/guides/gs/testing-web/
- https://quarkus.io/guides/getting-started-testing